content/hardware/05.pro-solutions/solutions-and-kits/edge-control/tutorials/smart-irrigation-system/content.md

```diff
@@ -0,0 +1,495 @@
+
+// Your OpenWeather API Key
+String openWeatherMapApiKey = "<API Key>";
+
+// Replace with your country code and city
+String city = "Santiago de los Caballeros";
+String countryCode = "DO";
+```
+
+Repetitively the MKR WiFi 1010 update its connection with the Arduino IoT Cloud, and checks for changes on every variable, from the valves switches to the scheduled activated timers. It shows when it's successfully connected to the cloud by turning on its blue LED. Also, it asks for a weather forecast every 10 minutes to update the temperature, humidity and rain probability in the Arduino IoT Cloud dashboard.
+
+```arduino
+void loop() {
+  // function that ask for schedule timers set on the Cloud
+  scheduleHandler();
+
+  unsigned long currentMillis = millis();
+
+  if (currentMillis - previousMillis >= intervalStat) {
+    // save the last time you asked for connectivity status
+    previousMillis = currentMillis;
+
+    // Turn on the blue LED if the board is successfully connected to the Cloud.
+    if (ArduinoCloud.connected()) {
+      WiFiDrv::analogWrite(27, 122);
+    } else {
+      WiFiDrv::analogWrite(27, 0);
+    }
+  }
+
+  // doing this just once after there's a WiFi connections
+  if (bootForecast == 1 && WiFi.status() == WL_CONNECTED) {
+    getForecast();  // request rain probability
+    getWeather();   // request temperature and humidity
+    bootForecast = 0;
+  }
+
+
+  if (millis() - lastRequest > interval) {
+    // request weather variables each interval
+    if (WiFi.status() == WL_CONNECTED) {
+      getForecast();
+      getWeather();
+    }
+    lastRequest = millis();
+  }
+
+  ArduinoCloud.update();
+}
+```
+
+The system needs to be as fast as possible when a valve activation is requested, this is why there is one function for each valve that is fired once the valve status is changed on the Arduino IoT Cloud.
+
+```arduino
+/*
+  Since Z1 is READ_WRITE variable, onZ1Change() is
+  executed every time a new value is received from IoT Cloud.
+*/
+void onZ1Change() {
+  // Add your code here to act upon Z1 change
+  Serial.println("Z1 changed!!");
+  SensorValues_t vals;
+  vals.z1_local = z1;
+  delay(1000);
+}
+```
+
+The `uploadValues(SensorValues_t *vals)` is executed every time the Edge Control requests data through I2C from the MKR. It updates all the local and Cloud shared variables to be sent, also it converts the water level variable to water volume using the cylinder volume formula `v = πr²h` where `r` is the tank radius in meters and `h` is the tank height in millimeters so the volume result is in liters.
+
+Using flow control variables we store the initial water level as a reference point, any level measured below that point will be considered as water use updating the `water_usage` variable, if we refill the tank, it will save the higher value of water level as a new reference point.
+
+```arduino
+  water_level = vals->water_level_local;
+  water_volume = PI * (radius) * (radius) * (int)water_level * 10;
```

Comment:
Here I would like to suggest a review of used measurement units within how it's processed inside the code, as I think I'm seeing something weird. Maybe it could use additional information if applicable to explain the unit measurements are processed adequately inside the code.

Reply:
I tried to explain it better and update the code expression. Let me know if it's clear enough now.